### DISCUSSION TITLE
add global menu

### DISCUSSION THREAD

**Discussion by SimoneRiccardi** [hyprwm/Hyprland#1358] [General]
It would be nice to have a global menu of the used application on the status bar like in macos or kde. 

It looks nicer and It's more space efficient without giving up features ( because you can remove the menu bar on top of all applications on the screen and show the menu options just of the application that you're using on the status bar)

**Comment by vaxerski** [comment]
not possible without a protocol

IIRC KDE already has one, but it's not a wayland standard and I'd refrain from implementing KDE protocols.

**Reply by SimoneRiccardi** [reply]
OK

Just a couple of question: 

1. Let's say that the protocol exists, would have more sense manage that with the window manager (like hyprland or sway) or with the bar (like Waybar or Hybrid)?  
2. For you who is the best candidate for inventing and implementing the protocol ? (xdg-desktop-portal? wayland? wlroots? someone else?)

**Reply by vaxerski** [reply]
1. probably the bar
2. wayland

**Reply by SimoneRiccardi** [reply]
Ok thank you

**Reply by lestcape** [reply]
This is not a good (in my opinion) or standard protocol:
https://wayland.app/protocols/kde-appmenu

This is what i think should really help in a good way for to have a cross toolkit/desktop global menu in a future:
https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/52

If you want to have a global menu, you should ask for the support of the dbus_annotation protocol. If it becomes a standard, it is likely that it can then be used to define the global menu and then use it between different toolkits and desktops. I think most environments have no objection to using it, but there is also no rush for it to be implemented and become a standard :(

**Reply by malikwirin** [reply]
@lestcape so https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/52 is the blocker at the moment for an official wayland standard? You know about the current state?

**Reply by lestcape** [reply]
If you want a cross desktop global menu for Wayland applications, then yes, this is the most close way to have it and the one that at least accepts GNOME. There are more variants, but the GNOME people don't like the rest. However, it's a long road. It needs to be merged and accepted by everyone involved in Wayland, then implemented in at least the main projects that use Wayland for the desktop. Furthermore, all toolkits must implement it (this will be especially difficult with GTK, and the longer it takes, the worse it will get). The application developer has to use what the toolkit makes available (in GNOME applications, for example, there aren't even menus to export), and when all that happens, then you'll be able to display the menus of a GTK application running on Wayland in KDE or in other desktop environment with the support for it. It's a long run... they've barely made preparations and some may get tired along the way.

**Comment by skela** [comment]
ive got a global menu in qtile. doesnt work for all apps though, but some, weirdly enough most if not all electron apps seem to use dbus stuff to do it. shame there's no protocol that covers all apps.

global menus make a lot of sense in tiling window managers imho, since you've gotten rid of the window decorations etc, but apps with a menu at the top, just look broken or out of place. Its one of the main things thats preventing me from using hyprland as a daily driver.

but i agree, makes more sense to implement this stuff in a Bar, hopefully Waybar will do it one day.

**Reply by Internetperson-dev** [reply]
https://hellosystem.github.io/docs/developer/menu.html Seems to document some of it, but it might be Xorg only? 

Maybe someone could try something prototyping something with hyprtavern if D-Bus protocols are not great.

**Comment by Kommynct** [comment]
https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/52 relevant protocol... but does involve dbus, sorry vaxry

**Reply by UjinT34** [reply]
Doesn't look like HL needs to deal with dbus itself. If I understand it correctly it only sets the names and types of dbus endpoints and compositor doesn't have to deal with them. Not very usable unless a bar itself supports relevant dbus interfaces but has a hard time linking specific endpoints to the clients (e.g. can't decide which endpoint to use with the currently focused window)

**Comment by edrozenberg** [comment]
A "fake" global menu might be a nice early step. A standard set of menus that apply to whichever is the active tile, and does standard tile things (sizing, moving, etc). This "fake" global menu would be the same for all tiles, would not include app-specific menu items, given the lack of a proper protocol atm. Benefit would be having some standard menu items always showing in the top left and looking more like a Mac :).

**Reply by malikwirin** [reply]
what you mean sounds more like window decorators than global menu
